python/uperf/uperf_plugin.py

```diff
@@ -0,0 +1,152 @@
+
+
+@dataclass
+class UPerfError:
+    """
+    This is the output data structure in the error  case.
+    """
+    error: str
+
+def start_client(params: UPerfParams):
+    process_env = os.environ.copy()
+    # Pass variables into profile.
+    process_env["h"] = params.server_addr
+    process_env["proto"] = params.protocol.value
+    process_env["dur"] = str(params.run_duration_ms) + "ms"
+    # TODO: Generate various types of profiles instead of using a sample profile.
+    # Note: uperf calls this 'master'
+    return subprocess.Popen(['uperf', '-vaR', '-i', '1', '-m', os.getcwd() + '/profiles/sample.xml', ],
+        stdout=subprocess.PIPE, env=process_env, cwd=os.getcwd())
+
+# TODO: Make running server its own step. And make it run on a timer.
+
+def process_output(output: bytes) -> typing.Tuple[str, typing.Union[UPerfResults, UPerfError]]:
+    decoded_output = output.decode("utf-8")
+    profile_run_search = re.search(r"running profile:(.+) \.\.\.", decoded_output)
+    if (profile_run_search == None):
+        return "error", UPerfError("Failed to parse output: could not find profile name.\nOutput: " + decoded_output)
+
+    profile_run = profile_run_search.group(1)
+
+    timeseries_data = {}
+
+    # There are multiple values for the name field. What we care about depends on the workload.
+    tx_name = "Txn2" # TODO: Account for other possible desired values.
+    timeseries_data_search = re.findall(rf"timestamp_ms:([\d\.]+) name:{tx_name} nr_bytes:(\d+) nr_ops:(\d+)", decoded_output)
+    for datapoint in timeseries_data_search:
+        # For now, multiplying by 1000 to get unique times as integers.
+        timeseries_data[int(float(datapoint[0])*1000)] = UPerfRawData(int(datapoint[1]), int(datapoint[2]))
+
+    if (len(timeseries_data_search) == 0):
+        return "error", UPerfError("No results found.\nOutput: " + decoded_output)
+
+    return "success", UPerfResults(profile_name=profile_run, raw=timeseries_data)
+
+# The following is a decorator (starting with @). We add this in front of our function to define the metadata for our
+# step.
+@plugin.step(
+    id="uperf",
+    name="UPerf Run",
+    description="Runs uperf locally",
+    outputs={"success": UPerfResults, "error": UPerfError},
+)
+def run_uperf(params: UPerfParams) -> typing.Tuple[str, typing.Union[UPerfResults, UPerfError]]:
+    """
+    Runs a uperf benchmark locally
+
+    :param params:
+
+    :return: the string identifying which output it is, as well the output structure
+    """
+    with start_client(params) as master_process:
+        outs, errs = master_process.communicate()
+
+    if errs != None and len(errs) > 0:
+        return "error", UPerfError(errs.decode("utf-8"))
+    if outs.find(b"aborted") != -1:
+        return "error", UPerfError("Errors found in run. Output:\n" + outs.decode("utf-8"))
+
+    # Debug output
+    print(outs.decode("utf-8"))
+
+    return process_output(outs)
+
+@plugin.step(
```

Comment:
Complete nit pick and functionally inconsequential, but I would put the server step above the client step because it aligns better with the workflow.